src/core/admittance_matrix.jl

```diff
@@ -0,0 +1,262 @@
+"""
+struct AdmittanceMatrix{T}
+    idx_to_bus::Vector{Int}
+    bus_to_idx::Dict{Int,Int}
+    matrix::T
+    # TODO should this include a RHS?
+end
+
+# Generic Case (e.g. with Dense Array Case)
+Base.show(io::IO, x::AdmittanceMatrix{<:Any}) = print(io, "AdmittanceMatrix($(length(x.idx_to_bus)) buses, $(length(x.matrix)) entries)")
+
+# Typical Sparse Array Case
+Base.show(io::IO, x::AdmittanceMatrix{<:SparseArrays.AbstractSparseMatrix}) = print(io, "AdmittanceMatrix($(length(x.idx_to_bus)) buses, $(length(nonzeros(x.matrix))) entries)")
+
+
+
+"""
+Stores metadata related to an PTDF Matirx
+
+Designed to work with the inverse of both complex (i.e. Y) and real-valued (e.g. b) valued
+admittance matrices.
+
+Typically the matrix will be dense.
+"""
+struct PowerTransferDistributionFactors{T}
+    idx_to_bus::Vector{Int}
+    bus_to_idx::Dict{Int,Int}
+    matrix::T
+    # TODO should this include a RHS?
+end
+
+# Generic Case (e.g. with Dense Array Case)
+Base.show(io::IO, x::PowerTransferDistributionFactors{<:Any}) = print(io, "PowerTransferDistributionFactors($(length(x.idx_to_bus)) buses, $(length(x.matrix)) entries)")
+
+
+
+
+"data should be a PowerModels network data model"
+function calc_susceptance_matrix(data::Dict{String,<:Any})
+    if length(data["dcline"]) > 0
+        Memento.error(_LOGGER, "calc_susceptance_matrix does not support data with dclines")
+    end
+    if length(data["switch"]) > 0
+        Memento.error(_LOGGER, "calc_susceptance_matrix does not support data with switches")
+    end
+
+    buses = [x.second for x in data["bus"] if (x.second[pm_component_status["bus"]] != pm_component_status_inactive["bus"])]
+    sort!(buses, by=x->x["index"])
+
+    idx_to_bus = [x["index"] for x in buses]
+    bus_to_idx = Dict(x["index"] => i for (i,x) in enumerate(buses))
+    #println(idx_to_bus)
+    #println(bus_to_idx)
+
+    I = Int64[]
+    J = Int64[]
+    V = Float64[]
+
+    for (i,branch) in data["branch"]
+        if branch[pm_component_status["branch"]] != pm_component_status_inactive["branch"]
+            f_bus = bus_to_idx[branch["f_bus"]]
+            t_bus = bus_to_idx[branch["t_bus"]]
+            b_val = -branch["br_x"]/(branch["br_x"]^2+branch["br_r"]^2)
+            push!(I, f_bus); push!(J, t_bus); push!(V,  b_val)
+            push!(I, t_bus); push!(J, f_bus); push!(V,  b_val)
+            push!(I, f_bus); push!(J, f_bus); push!(V, -b_val)
+            push!(I, t_bus); push!(J, t_bus); push!(V, -b_val)
+        end
+    end
+
+    m = sparse(I,J,V)
+    #println(m)
+
+    return AdmittanceMatrix(idx_to_bus, bus_to_idx, m)
+end
+
+"note, data should be a PowerModels network data model"
+function calc_ptdf_matrix(data::Dict{String,<:Any})
+    #TODO check single connected component
+
+    ref_bus = reference_bus(data)
+    sm = calc_susceptance_matrix(data)
+    return calc_ptdf_matrix(sm, sm.bus_to_idx[ref_bus["index"]])
+end
+
+"calculates a PTDF matrix"
```

Comment:
Should add to the documentation this doesn't consider distributed slack.

Reply:
I am not really sure what you mean by distributed slack at this line of the code.

In the `ptdf_opf` formulation no slack bus is required, e.g. see https://github.com/lanl-ansi/PowerModels.jl/blob/ptdf/src/prob/opf.jl#L160

In the part of the new code that assumes a single slack bus is the dc_pf solver,

https://github.com/lanl-ansi/PowerModels.jl/blob/ptdf/src/core/admittance_matrix.jl#L210

This follows the same convention as the JuMP version of the model, so maybe your suggestion would be to add a note to the PowerModels PF docs that they assume a single slack bus?

Reply:
I'll also add, that I make a subtle distinction between `reference_bus` and `slack_bus`.  For me, `reference_bus` means that theta is fixed at a bus and `slack_bus` means that the power injection at a bus is a free variable.

Reply:
Depending on how you plan to use the PTDF the slack buses need to be considered. This is the Matpower documentation https://github.com/MATPOWER/matpower/blob/master/lib/makePTDF.m and we do the same in PowerSystems.jl https://github.com/NREL/PowerSystems.jl/blob/master/src/utils/network_calculations/ptdf_calculations.jl 

If the PTDF will be used for SCOPF or reliability analysis it needs to be taken into account. I know that this is not the intended use for now. My suggestion is just to mention that the PTDF making code doesn't this into account yet. 

Finally, the distinction between reference and slack in this context.

Reply:
Thanks for these links.  I am wondering if the disconnect here goes back to my point of not really implementing the traditional PDTF here.  In both of these links the PDTF matrix is nbr-by-nb, while in my implementation all I consider is the inverse of the admittance matrix (i.e. nb-by-nb).

This admittance matrix inverse fixes a refrence bus but is agnostic to how the nodal injections are balanced.  At the time of setting up the PF problem one can select any type of balancing function, including a single free nodal injection variable, a linear function of nodal injection variables, free nodal injection variables and an objective function.  I suppose that's why I designed it this way, so that one can delay making any choices about the power balance until later.